crates/bevy_utils/src/intern.rs

```diff
@@ -0,0 +1,306 @@
+//! Provides types used to statically intern immutable values.
+//!
+//! Interning is a pattern used to save memory by deduplicating identical values,
+//! speed up code by shrinking the stack size of large types,
+//! and make comparisons for any type as fast as integers.
+
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    ops::Deref,
+    ptr,
+    sync::{OnceLock, PoisonError, RwLock},
+};
+
+use crate::HashSet;
+
+/// An interned value. Will stay valid until the end of the program and will not drop.
+///
+/// For details on interning, see [the module level docs](self).
+///
+/// # Comparisons
+///
+/// Interned values use reference equality, meaning they implement [`Eq`]
+/// and [`Hash`] regardless of whether `T` implements these traits.
+/// Two interned values will only compare equal if they were interned using
```

Comment:
```suggestion
/// Two interned values are only guaranteed to compare equal if they were interned using
```

A slight correction to this phrasing I wrote. Type that implement `static_ref` will compare equal regardless of the interner used.